In [78]:
#import necessary libraries
import numpy as np
import pandas as pd

In [79]:
#read the dataset
tc = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv',index_col=0)

In [80]:
#display the dataset
tc.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
customerID                                                                  
7590-VHVEG  Female              0     Yes         No       1           No   
5575-GNVDE    Male              0      No         No      34          Yes   
3668-QPYBK    Male              0      No         No       2          Yes   
7795-CFOCW    Male              0      No         No      45           No   
9237-HQITU  Female              0      No         No       2          Yes   

               MultipleLines InternetService OnlineSecurity OnlineBackup  \
customerID                                                                 
7590-VHVEG  No phone service             DSL             No          Yes   
5575-GNVDE                No             DSL            Yes           No   
3668-QPYBK                No             DSL            Yes          Yes   
7795-CFOCW  No phone service             DSL            Yes           No   
9237-HQITU                No     Fiber optic             No           No   

           DeviceProtection TechSupport StreamingTV StreamingMovies  \
customerID                                                            
7590-VHVEG               No          No          No              No   
5575-GNVDE              Yes          No          No              No   
3668-QPYBK               No          No          No              No   
7795-CFOCW              Yes         Yes          No              No   
9237-HQITU               No          No          No              No   

                  Contract PaperlessBilling              PaymentMethod  \
customerID                                                               
7590-VHVEG  Month-to-month              Yes           Electronic check   
5575-GNVDE        One year               No               Mailed check   
3668-QPYBK  Month-to-month              Yes               Mailed check   
7795-CFOCW        One year               No  Bank transfer (automatic)   
9237-HQITU  Month-to-month              Yes           Electronic check   

            MonthlyCharges TotalCharges Churn  
customerID                                     
7590-VHVEG           29.85        29.85    No  
5575-GNVDE           56.95       1889.5    No  
3668-QPYBK           53.85       108.15   Yes  
7795-CFOCW           42.30      1840.75    No  
9237-HQITU           70.70       151.65   Yes

In [81]:
#get concise summary of dataset
tc.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [82]:
#find if missing values are present
tc.isnull().sum()

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [83]:
#no missing values, dataset is clean
tc.shape

(7043, 20)

In [84]:
#making a dataframe of males
male = tc.loc[tc.gender=='Male']
male.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
customerID                                                                 
5575-GNVDE   Male              0      No         No      34          Yes   
3668-QPYBK   Male              0      No         No       2          Yes   
7795-CFOCW   Male              0      No         No      45           No   
1452-KIOVK   Male              0      No        Yes      22          Yes   
6388-TABGU   Male              0      No        Yes      62          Yes   

               MultipleLines InternetService OnlineSecurity OnlineBackup  \
customerID                                                                 
5575-GNVDE                No             DSL            Yes           No   
3668-QPYBK                No             DSL            Yes          Yes   
7795-CFOCW  No phone service             DSL            Yes           No   
1452-KIOVK               Yes     Fiber optic             No          Yes   
6388-TABGU                No             DSL            Yes          Yes   

           DeviceProtection TechSupport StreamingTV StreamingMovies  \
customerID                                                            
5575-GNVDE              Yes          No          No              No   
3668-QPYBK               No          No          No              No   
7795-CFOCW              Yes         Yes          No              No   
1452-KIOVK               No          No         Yes              No   
6388-TABGU               No          No          No              No   

                  Contract PaperlessBilling              PaymentMethod  \
customerID                                                               
5575-GNVDE        One year               No               Mailed check   
3668-QPYBK  Month-to-month              Yes               Mailed check   
7795-CFOCW        One year               No  Bank transfer (automatic)   
1452-KIOVK  Month-to-month              Yes    Credit card (automatic)   
6388-TABGU        One year               No  Bank transfer (automatic)   

            MonthlyCharges TotalCharges Churn  
customerID                                     
5575-GNVDE           56.95       1889.5    No  
3668-QPYBK           53.85       108.15   Yes  
7795-CFOCW           42.30      1840.75    No  
1452-KIOVK           89.10       1949.4    No  
6388-TABGU           56.15      3487.95    No

Number of males = 3555

In [85]:
#making a dataframe of females
female = tc.loc[tc.gender=='Female']
female.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
customerID                                                                  
7590-VHVEG  Female              0     Yes         No       1           No   
9237-HQITU  Female              0      No         No       2          Yes   
9305-CDSKC  Female              0      No         No       8          Yes   
6713-OKOMC  Female              0      No         No      10           No   
7892-POOKP  Female              0     Yes         No      28          Yes   

               MultipleLines InternetService OnlineSecurity OnlineBackup  \
customerID                                                                 
7590-VHVEG  No phone service             DSL             No          Yes   
9237-HQITU                No     Fiber optic             No           No   
9305-CDSKC               Yes     Fiber optic             No           No   
6713-OKOMC  No phone service             DSL            Yes           No   
7892-POOKP               Yes     Fiber optic             No           No   

           DeviceProtection TechSupport StreamingTV StreamingMovies  \
customerID                                                            
7590-VHVEG               No          No          No              No   
9237-HQITU               No          No          No              No   
9305-CDSKC              Yes          No         Yes             Yes   
6713-OKOMC               No          No          No              No   
7892-POOKP              Yes         Yes         Yes             Yes   

                  Contract PaperlessBilling     PaymentMethod  MonthlyCharges  \
customerID                                                                      
7590-VHVEG  Month-to-month              Yes  Electronic check           29.85   
9237-HQITU  Month-to-month              Yes  Electronic check           70.70   
9305-CDSKC  Month-to-month              Yes  Electronic check           99.65   
6713-OKOMC  Month-to-month               No      Mailed check           29.75   
7892-POOKP  Month-to-month              Yes  Electronic check          104.80   

           TotalCharges Churn  
customerID                     
7590-VHVEG        29.85    No  
9237-HQITU       151.65   Yes  
9305-CDSKC        820.5   Yes  
6713-OKOMC        301.9    No  
7892-POOKP      3046.05   Yes

Number of Females = 3488
There are almost same number of male and female customers

In [86]:
#making a dataframe of Senior Citizens
sc = tc.loc[tc.SeniorCitizen==1]
sc.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
customerID                                                                  
8779-QRDMV    Male              1      No         No       1           No   
3841-NFECX  Female              1     Yes         No      71          Yes   
4929-XIHVW    Male              1     Yes         No       2          Yes   
3413-BMNZE    Male              1      No         No       1          Yes   
8012-SOUDQ  Female              1      No         No      43          Yes   

               MultipleLines InternetService OnlineSecurity OnlineBackup  \
customerID                                                                 
8779-QRDMV  No phone service             DSL             No           No   
3841-NFECX               Yes     Fiber optic            Yes          Yes   
4929-XIHVW                No     Fiber optic             No           No   
3413-BMNZE                No             DSL             No           No   
8012-SOUDQ               Yes     Fiber optic             No          Yes   

           DeviceProtection TechSupport StreamingTV StreamingMovies  \
customerID                                                            
8779-QRDMV              Yes          No          No             Yes   
3841-NFECX              Yes         Yes          No              No   
4929-XIHVW              Yes          No         Yes             Yes   
3413-BMNZE               No          No          No              No   
8012-SOUDQ               No          No         Yes              No   

                  Contract PaperlessBilling              PaymentMethod  \
customerID                                                               
8779-QRDMV  Month-to-month              Yes           Electronic check   
3841-NFECX        Two year              Yes    Credit card (automatic)   
4929-XIHVW  Month-to-month              Yes    Credit card (automatic)   
3413-BMNZE  Month-to-month               No  Bank transfer (automatic)   
8012-SOUDQ  Month-to-month              Yes           Electronic check   

            MonthlyCharges TotalCharges Churn  
customerID                                     
8779-QRDMV           39.65        39.65   Yes  
3841-NFECX           96.35      6766.95    No  
4929-XIHVW           95.50       181.65    No  
3413-BMNZE           45.25        45.25    No  
8012-SOUDQ           90.25      3838.75    No

Number of Senior Citizens = 1142

In [87]:
#making a dataframe of Non-Senior Citizens
nsc = tc.loc[tc.SeniorCitizen==0]
nsc.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
customerID                                                                  
7590-VHVEG  Female              0     Yes         No       1           No   
5575-GNVDE    Male              0      No         No      34          Yes   
3668-QPYBK    Male              0      No         No       2          Yes   
7795-CFOCW    Male              0      No         No      45           No   
9237-HQITU  Female              0      No         No       2          Yes   

               MultipleLines InternetService OnlineSecurity OnlineBackup  \
customerID                                                                 
7590-VHVEG  No phone service             DSL             No          Yes   
5575-GNVDE                No             DSL            Yes           No   
3668-QPYBK                No             DSL            Yes          Yes   
7795-CFOCW  No phone service             DSL            Yes           No   
9237-HQITU                No     Fiber optic             No           No   

           DeviceProtection TechSupport StreamingTV StreamingMovies  \
customerID                                                            
7590-VHVEG               No          No          No              No   
5575-GNVDE              Yes          No          No              No   
3668-QPYBK               No          No          No              No   
7795-CFOCW              Yes         Yes          No              No   
9237-HQITU               No          No          No              No   

                  Contract PaperlessBilling              PaymentMethod  \
customerID                                                               
7590-VHVEG  Month-to-month              Yes           Electronic check   
5575-GNVDE        One year               No               Mailed check   
3668-QPYBK  Month-to-month              Yes               Mailed check   
7795-CFOCW        One year               No  Bank transfer (automatic)   
9237-HQITU  Month-to-month              Yes           Electronic check   

            MonthlyCharges TotalCharges Churn  
customerID                                     
7590-VHVEG           29.85        29.85    No  
5575-GNVDE           56.95       1889.5    No  
3668-QPYBK           53.85       108.15   Yes  
7795-CFOCW           42.30      1840.75    No  
9237-HQITU           70.70       151.65   Yes

Number of Non seniors = 5901
There are almost 5 times Non Seniors than Seniors

In [88]:
#calculating monthly averages
nsc_mean = nsc.MonthlyCharges.mean()
sc_mean = sc.MonthlyCharges.mean()
print(nsc_mean,sc_mean)

61.84744111167598 79.82035901926453


Monthly Charges for Senior Citizens is more

In [89]:
#finding out number of customers having phone service
nsc_ps = nsc.PhoneService.map(lambda desc: "Yes" in desc).sum()
sc_ps = sc.PhoneService.map(lambda desc: "Yes" in desc).sum()
print(nsc_ps,sc_ps)

5323 1038


In [90]:
nsc_psx = (nsc_ps/len(nsc))*100
sc_psx = (sc_ps/len(sc))*100
print(nsc_psx,sc_psx)

90.20504999152686 90.89316987740806


In both cases, almost 90% of the customers have phone service

In [91]:
#finding out number of people not having internet service
nsc_nis = nsc.InternetService.map(lambda desc: "No" in desc).sum()
sc_nis = sc.InternetService.map(lambda desc: "No" in desc).sum()
print(nsc_nis,sc_nis)

1474 52


In [92]:
nsc_nisx = (nsc_nis/len(nsc))*100
sc_nisx = (sc_nis/len(sc))*100
print(nsc_nisx,sc_nisx)

24.978817149635653 4.553415061295972


So, around 25% of Non Seniors do not have Internet Connections
and only 4% of Seniors do not have Internet Connections

In [93]:
#finding the type of internet connections
nsc_isdsl = nsc.InternetService.map(lambda desc: "DSL" in desc).sum()
nsc_fo = len(nsc) - nsc_isdsl - nsc_nis
sc_isdsl = sc.InternetService.map(lambda desc: "DSL" in desc).sum()
sc_fo = len(sc) - sc_isdsl - sc_nis
print(nsc_isdsl,sc_isdsl)
print(nsc_fo,sc_fo)

2162 259
2265 831


In [94]:
nsc_fox = (nsc_fo/len(nsc))*100
sc_fox = (sc_fo/len(sc))*100
print(nsc_fox,sc_fox)

38.383324860193184 72.76707530647985


Around 73% of Seniors  and 38% of non-seniors have fibre optic connections.

In [95]:
#Mean monthly charges for non-seniors with internet
nscis = nsc.loc[nsc.InternetService!='No']
nscis.MonthlyCharges.mean()

75.43117235147945

In [96]:
#Mean monthly charges for seniors with internet
scis = sc.loc[sc.InternetService!='No']
scis.MonthlyCharges.mean()

82.58344036697254

Mean monthly charges for seniors is more than non-seniors with internet connection

In [97]:
#calculating the churn amount for the two types of users
nsc_churn = nsc.Churn.map(lambda desc: "Yes" in desc).sum()
sc_churn = sc.Churn.map(lambda desc: "Yes" in desc).sum()
print(nsc_churn,sc_churn)

1393 476


In [98]:
nsc_churnx = (nsc_churn/len(nsc))*100
sc_churnx = (sc_churn/len(sc))*100
print(nsc_churnx,sc_churnx)

23.606168446026096 41.68126094570928


The amount for churn for seniors is more.(~41%),i.e., when considering only Senior Citizen's data, there is more tendency for them to churn when compared to non-seniors.

In [99]:
#finding churn amt. for users with internet
nsc_churn = nscis.Churn.map(lambda desc: "Yes" in desc).sum()
sc_churn = scis.Churn.map(lambda desc: "Yes" in desc).sum()
print(nsc_churn,sc_churn)

1285 471


In [100]:
nsc_churnx = (nsc_churn/len(nscis))*100
sc_churnx = (sc_churn/len(scis))*100
print(nsc_churnx,sc_churnx)

29.026428732776143 43.211009174311926


Almost same portion of churn for seniors but for non-seniors the amount has increased by 6% 

In [101]:
#plotting tables to obtain relationship between Senior Citizen and Churn
pd.crosstab(index = tc['SeniorCitizen'],columns=tc['Churn'],normalize=True)

Churn                No       Yes
SeniorCitizen                    
0              0.640068  0.197785
1              0.094562  0.067585

When we look the data as a whole, we see that the people who are not seniors churn more with percentage of 19%. This is different from the earlier case, wherein we divided the dataset into two and then analysed it.

In [102]:
#plotting tables to obtain relationship between Senior Citizen and Churn(conditional prob.)
pd.crosstab(index = tc['SeniorCitizen'],columns=tc['Churn'],normalize='index')

Churn                No       Yes
SeniorCitizen                    
0              0.763938  0.236062
1              0.583187  0.416813

Here,as seen before percentage of seniors churning is 41% much more than non-seniors(23%). Here, we looked at the probabiblity that a given a person will churn provided the condition that he/she is a senior or not.

In [103]:
#obtaining new dataset of only numeric types
num_data = tc.select_dtypes(exclude=[object])
num_data.head()

SeniorCitizen  tenure  MonthlyCharges
customerID                                       
7590-VHVEG              0       1           29.85
5575-GNVDE              0      34           56.95
3668-QPYBK              0       2           53.85
7795-CFOCW              0      45           42.30
9237-HQITU              0       2           70.70

In [104]:
#finding Pearson correlation
pearson_mat = num_data.corr()
pearson_mat

SeniorCitizen    tenure  MonthlyCharges
SeniorCitizen        1.000000  0.016567        0.220173
tenure               0.016567  1.000000        0.247900
MonthlyCharges       0.220173  0.247900        1.000000

In [105]:
#finding Spearman's correlation
spearman_mat = tc.corr(method='spearman')
spearman_mat

SeniorCitizen    tenure  MonthlyCharges
SeniorCitizen        1.000000  0.018556        0.221067
tenure               0.018556  1.000000        0.276417
MonthlyCharges       0.221067  0.276417        1.000000

In [106]:
#finding Kendall's correlation
kendall_mat = num_data.corr(method='kendall')
kendall_mat

SeniorCitizen    tenure  MonthlyCharges
SeniorCitizen        1.000000  0.015315        0.180636
tenure               0.015315  1.000000        0.190565
MonthlyCharges       0.180636  0.190565        1.000000

From all the above three correaltions, we see that none of the factors are related to each other very much. What we want to see is whether these factors affect the 'Churn' outcome:

In [107]:
Churn = {'No':0,'Yes':1}

In [108]:
#replacing values of Churn attributes
pd.set_option('mode.chained_assignment', None)
tc.Churn[tc.Churn == 'Yes'] = 1
tc.Churn[tc.Churn == 'No'] = 0
tc.Contract[tc.Contract == 'Month-to-month'] = 0
tc.Contract[tc.Contract == 'One year'] = 1
tc.Contract[tc.Contract == 'Two year'] = 2

In [109]:
#Converting the data type to int
tc['Churn'] = tc['Churn'].astype(int)
tc['Contract'] = tc['Contract'].astype(int)

In [110]:
tc.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract              int32
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                 int32
dtype: object

In [111]:
num_data = tc.select_dtypes(exclude=[object])
num_data.head()

SeniorCitizen  tenure  Contract  MonthlyCharges  Churn
customerID                                                        
7590-VHVEG              0       1         0           29.85      0
5575-GNVDE              0      34         1           56.95      0
3668-QPYBK              0       2         0           53.85      1
7795-CFOCW              0      45         1           42.30      0
9237-HQITU              0       2         0           70.70      1

In [112]:
#Now, obtaining the correlation
pearson_mat = num_data.corr()
pearson_mat

SeniorCitizen    tenure  Contract  MonthlyCharges     Churn
SeniorCitizen        1.000000  0.016567 -0.142554        0.220173  0.150889
tenure               0.016567  1.000000  0.671607        0.247900 -0.352229
Contract            -0.142554  0.671607  1.000000       -0.074195 -0.396713
MonthlyCharges       0.220173  0.247900 -0.074195        1.000000  0.193356
Churn                0.150889 -0.352229 -0.396713        0.193356  1.000000